cpu_scheduling_algorithms/non_preemptive_sjf_scheduling.cpp

```diff
@@ -0,0 +1,318 @@
+ * @tparam S Data type of Process ID
+ * @tparam T Data type of Arrival time
+ * @tparam E Data type of Burst time
+ */
+template <typename S, typename T, typename E>
+class Compare {
+ public:
+    /**
+     * @param t1 First tuple
+     * @param t2 Second tuple
+     * @brief A comparator function that checks whether to swap the two tuples
+     * or not.
+     * <a
+     * href="https://www.geeksforgeeks.org/comparator-class-in-c-with-examples/">
+     * detailed description of comparator </a>
+     * @returns true if the tuples SHOULD be swapped
+     * @returns false if the tuples SHOULDN'T be swapped
+     */
+    bool operator()(tuple<S, T, E, double, double, double>& t1,
+                    tuple<S, T, E, double, double, double>& t2) {
+        // Compare burst times for SJF
+        if (get<2>(t2) < get<2>(t1)) {
+            return true;
+        }
+        // If burst times are the same, compare arrival times
+        else if (get<2>(t2) == get<2>(t1)) {
+            return get<1>(t2) < get<1>(t1);
+        }
+        return false;
+    }
+};
+
+/**
+ * @class SJF
+ * @brief Class which implements the SJF scheduling algorithm
+ * @tparam S Data type of Process ID
+ * @tparam T Data type of Arrival time
+ * @tparam E Data type of Burst time
+ */
+template <typename S, typename T, typename E>
+class SJF {
+    /**
+     * Priority queue of schedules(stored as tuples) of processes.
+     * In each tuple
+     * 1st element: Process ID
+     * 2nd element: Arrival Time
+     * 3rd element: Burst time
+     * 4th element: Completion time
+     * 5th element: Turnaround time
+     * 6th element: Waiting time
+     */
+    priority_queue<tuple<S, T, E, double, double, double>,
+                   vector<tuple<S, T, E, double, double, double>>,
+                   Compare<S, T, E>>
+        schedule;
+
+    // Stores final status of all the processes after completing the execution.
+    vector<tuple<S, T, E, double, double, double>> result;
+
+    // Stores process IDs. Used for confirming absence of a process while  it.
+    unordered_set<S> idList;
+
+ public:
+    /**
+     * @brief Adds the process to the ready queue if it isn't already there
+     * @param id Process ID
+     * @param arrival Arrival time of the process
+     * @param burst Burst time of the process
+     * @returns void
+     *
+     */
+    void addProcess(S id, T arrival, E burst) {
+        // Add if a process with process ID as id is not found in idList.
+        if (idList.find(id) == idList.end()) {
+            tuple<S, T, E, double, double, double> t =
+                make_tuple(id, arrival, burst, 0, 0, 0);
+            schedule.push(t);
+            idList.insert(id);
+        }
+    }
+
+    /**
+   * @brief Algorithm for scheduling CPU processes according to the Shortest Job
+   First (SJF) scheduling algorithm.
+   *
+   * @details Non pre-emptive SJF is an algorithm that schedules processes based
+   on the length
```

Comment:
this special comment isnt formatted properly

Reply:
the comment has been formatted